flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/SingleCheckpointBarrierHandler.java

```diff
@@ -235,6 +257,23 @@ int getNumOpenChannels() {
 		return numOpenChannels;
 	}
 
+	private CheckpointBarrier maybeTimeout(CheckpointBarrier barrier) {
```

Comment:
Could we move the timeout logic into `AlternatingController`? It would feel more natural if such logic is there. But I'm also not sure if you have chosen it here because you have some ideas for active timeout already.

Reply:
I had to do it anyway, as after a couple of hours of debugging I realise, that if we are timeouting barriers on this high level from aligned to unaligned, we also need to `resumeConsumption` of the said channel...

Note, I need to convert the checkpoint barrier to unaligned inside `SingleChecpkointBarrierHandler`, as this is the class that's calling `notifyCheckpoint(checkpointBarrier)`, and the `checkpointBarrier` that we are passing in that callback must be converted to unaligned if (so that the task can do the unaligned checkpoint). So we can not move the whole `maybeTimeOut` logic entirely to the controllers - it needs to be called in `SingleChecpkointBarrierHandler`.

All in all, please check the fixup commit that I will publish later with the fix.